File: typescript/src/storage/LocalStorage.ts
import { StorageProvider } from "./StorageProvider";
import { FileMetadata } from "../types";
import fs from "fs-extra";
import path from "path";
import os from "os";

/**
 * Options for local file system storage
 */
export interface LocalStorageOptions {
  /**
   * Path where browser profiles will be stored
   */
  storagePath?: string;
}

/**
 * Local file system storage provider implementation
 */
export class LocalStorage implements StorageProvider {
  private basePath: string;

  constructor(storagePath?: string) {
    // Use provided path or default to ~/.browserstate
    this.basePath = storagePath || path.join(os.homedir(), ".browserstate");
    fs.ensureDirSync(this.basePath);
  }

  /**
   * Get the full path for a user
   */
  private getUserPath(userId: string): string {
    return path.join(this.basePath, userId);
  }

  /**
   * Get the full path for a session
   */
  private getSessionPath(userId: string, sessionId: string): string {
    return path.join(this.getUserPath(userId), sessionId);
  }

  /**
   * Downloads a browser session to a local directory
   */
  async download(userId: string, sessionId: string): Promise<string> {
    const sourcePath = this.getSessionPath(userId, sessionId);
    const targetPath = path.join(
      os.tmpdir(),
      "browserstate",
      userId,
      sessionId,
    );

    // Clear target directory if it exists
    await fs.emptyDir(targetPath);

    try {
      // Copy the session directory
      await fs.copy(sourcePath, targetPath);
      return targetPath;
    } catch (error: unknown) {
      // Ensure directory exists even if copy fails
      await fs.ensureDir(targetPath);
      const errorMessage =
        error instanceof Error ? error.message : String(error);
      console.error("Error downloading from local storage:", errorMessage);
      return targetPath;
    }
  }

  /**
   * Uploads a browser session from a local directory
   */
  async upload(
    userId: string,
    sessionId: string,
    filePath: string,
  ): Promise<void> {
    const targetPath = this.getSessionPath(userId, sessionId);

    try {
      // Ensure target directory exists
      await fs.ensureDir(targetPath);

      // Copy the session directory
      await fs.copy(filePath, targetPath);
    } catch (error: unknown) {
      const errorMessage =
        error instanceof Error ? error.message : String(error);
      console.error("Error uploading to local storage:", errorMessage);
      throw new Error(
        `Failed to upload session to local storage: ${errorMessage}`,
      );
    }
  }

  /**
   * Lists available sessions for a user
   */
  async listSessions(userId: string): Promise<string[]> {
    const userPath = this.getUserPath(userId);

    try {
      // Ensure user directory exists
      await fs.ensureDir(userPath);

      // List directories in user path
      const entries = await fs.readdir(userPath, { withFileTypes: true });
      return entries
        .filter((entry) => entry.isDirectory())
        .map((entry) => entry.name);
    } catch (error: unknown) {
      const errorMessage =
        error instanceof Error ? error.message : String(error);
      console.error("Error listing sessions from local storage:", errorMessage);
      return [];
    }
  }

  /**
   * Deletes a browser session
   */
  async deleteSession(userId: string, sessionId: string): Promise<void> {
    const sessionPath = this.getSessionPath(userId, sessionId);

    try {
      if (await fs.pathExists(sessionPath)) {
        await fs.remove(sessionPath);
      }
    } catch (error: unknown) {
      const errorMessage =
        error instanceof Error ? error.message : String(error);
      console.error("Error deleting session from local storage:", errorMessage);
      throw new Error(
        `Failed to delete session from local storage: ${errorMessage}`,
      );
    }
  }

  /**
   * Downloads a single file from storage
   */
  async downloadFile(s3Key: string, localPath: string): Promise<boolean> {
    try {
      const sourcePath = path.join(this.basePath, s3Key);
      if (!(await fs.pathExists(sourcePath))) {
        return false;
      }
      await fs.copy(sourcePath, localPath);
      return true;
    } catch (error) {
      console.error(`Error downloading file from local storage:`, error);
      return false;
    }
  }

  /**
   * Uploads a single file to storage
   */
  async uploadFile(filePath: string, s3Key: string): Promise<void> {
    try {
      const targetPath = path.join(this.basePath, s3Key);
      await fs.ensureDir(path.dirname(targetPath));
      await fs.copy(filePath, targetPath);
    } catch (error) {
      console.error(`Error uploading file to local storage:`, error);
      throw new Error(`Failed to upload file to local storage: ${error}`);
    }
  }

  /**
   * Gets metadata for a session
   */
  async getMetadata(
    userId: string,
    sessionId: string,
  ): Promise<Map<string, FileMetadata>> {
    const sessionPath = this.getSessionPath(userId, sessionId);
    const metadataPath = path.join(sessionPath, ".browserstate-metadata.json");

    try {
      if (await fs.pathExists(metadataPath)) {
        const metadataContent = await fs.readFile(metadataPath, "utf8");
        const metadataObj = JSON.parse(metadataContent);
        return new Map(Object.entries(metadataObj));
      }
      return new Map();
    } catch (error) {
      console.error(`Error loading metadata from local storage:`, error);
      return new Map();
    }
  }

  /**
   * Saves metadata for a session
   */
  async saveMetadata(
    userId: string,
    sessionId: string,
    metadata: Map<string, FileMetadata>,
  ): Promise<void> {
    const sessionPath = this.getSessionPath(userId, sessionId);
    const metadataPath = path.join(sessionPath, ".browserstate-metadata.json");

    try {
      await fs.ensureDir(sessionPath);
      const metadataObj = Object.fromEntries(metadata);
      await fs.writeJSON(metadataPath, metadataObj, { spaces: 2 });
    } catch (error) {
      console.error(`Error saving metadata to local storage:`, error);
      throw new Error(`Failed to save metadata to local storage: ${error}`);
    }
  }
}

File: python/browserstate/storage/local_storage.py
import os
import shutil
from pathlib import Path
from typing import List, Optional
import tempfile
import asyncio

from .storage_provider import StorageProvider

class LocalStorage(StorageProvider):
    """
    Storage provider implementation that uses the local file system.
    """
    
    def __init__(self, storage_path: Optional[str] = None):
        """
        Initialize a LocalStorage provider.
        
        Args:
            storage_path: Path where browser profiles will be stored. Defaults to ~/.browserstate
        """
        self.base_path = storage_path or os.path.join(os.path.expanduser("~"), ".browserstate")
        # Ensure base directory exists
        os.makedirs(self.base_path, exist_ok=True)
    
    def _get_user_path(self, user_id: str) -> str:
        """
        Get path for a specific user's data.
        
        Args:
            user_id: User identifier
            
        Returns:
            Full path to the user's data directory
        """
        user_path = os.path.join(self.base_path, user_id)
        os.makedirs(user_path, exist_ok=True)
        return user_path
    
    def _get_session_path(self, user_id: str, session_id: str) -> str:
        """
        Get path for a specific session.
        
        Args:
            user_id: User identifier
            session_id: Session identifier
            
        Returns:
            Full path to the session directory
        """
        return os.path.join(self._get_user_path(user_id), session_id)
    
    def _get_temp_path(self, user_id: str, session_id: str) -> str:
        """
        Get a temporary path for a session.

        Args:
            user_id: User identifier
            session_id: Session identifier

        Returns:
            Full path to the temporary session directory
        """
        
        temp_dir = os.path.join(tempfile.gettempdir(), "browserstate", user_id)
        os.makedirs(temp_dir, exist_ok=True)
        return os.path.join(temp_dir, session_id)
    
    async def download(self, user_id: str, session_id: str) -> str:
        """
        Downloads a browser session to local temp directory.
        
        Args:
            user_id: User identifier
            session_id: Session identifier
            
        Returns:
            Path to the local directory containing the session data
        """
        session_path = self._get_session_path(user_id, session_id)
        target_path = self._get_temp_path(user_id, session_id)
        
        # Check if session exists
        if os.path.exists(session_path):
            # Clear target directory if it already exists
            if os.path.exists(target_path):
                shutil.rmtree(target_path)
            
            # Create target directory
            os.makedirs(target_path, exist_ok=True)
            
            # Run copy in executor to avoid blocking
            await asyncio.get_event_loop().run_in_executor(
                None,
                shutil.copytree,
                session_path,
                target_path,
                True  # dirs_exist_ok
            )
        else:
            # Create an empty directory for new sessions
            os.makedirs(target_path, exist_ok=True)
        
        return target_path
    
    async def upload(self, user_id: str, session_id: str, file_path: str) -> None:
        """
        Uploads browser session files from temp to storage.
        
        Args:
            user_id: User identifier
            session_id: Session identifier
            file_path: Path to the local directory containing session data
        """
        session_path = self._get_session_path(user_id, session_id)
        
        # Ensure session directory exists
        os.makedirs(session_path, exist_ok=True)
        
        # Run copy in executor to avoid blocking
        await asyncio.get_event_loop().run_in_executor(
            None,
            shutil.copytree,
            file_path,
            session_path,
            True  # dirs_exist_ok
        )
    
    async def list_sessions(self, user_id: str) -> List[str]:
        """
        Lists all available sessions for a user.
        
        Args:
            user_id: User identifier
            
        Returns:
            List of session identifiers
        """
        user_path = self._get_user_path(user_id)
        
        try:
            # Run scandir in executor to avoid blocking
            entries = await asyncio.get_event_loop().run_in_executor(
                None,
                os.scandir,
                user_path
            )
            return [entry.name for entry in entries if entry.is_dir()]
        except OSError:
            # Directory does not exist, return empty array
            return []
    
    async def delete_session(self, user_id: str, session_id: str) -> None:
        """
        Deletes a session.
        
        Args:
            user_id: User identifier
            session_id: Session identifier
        """
        session_path = self._get_session_path(user_id, session_id)
        
        if os.path.exists(session_path):
            # Run rmtree in executor to avoid blocking
            await asyncio.get_event_loop().run_in_executor(
                None,
                shutil.rmtree,
                session_path
            ) 